text/0000-pi-types.md

```diff
@@ -0,0 +1,301 @@
+      Π ⊢ const x: T
+      ──────────────
+      Π ⊢ x: T
+
+In human language, this simply means that one can use a constant parameter,
+`const x: T`, in expression context, as a value of type `T`.
+
+On the type level, we use the very same semantics as the ones generic
+parameters currently follows.
+
+## `const fn`s as Π-constructors
+
+We are interested in value dependency, but at the same time, we want to avoid
+complications such as [SMT-solvers](https://en.wikipedia.org/wiki/Satisfiability_modulo_theories).
+
+We achieve this by `const fn`, which allows us to take some const parameter and
+map it by some arbitrary, pure function, following the rules described in [RFC
+0911](https://github.com/rust-lang/rfcs/blob/master/text/0911-const-fn.md#detailed-design).
+
+## Type inference
+
+Since we are able to evaluate the function at compile time, we can easily infer
+const parameters, by adding an unification relation, simply
+
+    PiRelationInference
+      Γ ⊢ y = f(x)
+      Γ ⊢ T: U<y>
+      ──────────────
+      Γ ⊢ T: U<f(x)>
+
+Informally, this means that, you can substitute equal terms (in this case, `const fn` relations).
+
+The relational edge between two const parameters is simple a const fn, which is
+resolved under unification.
+
+We add an extra rule to improve inference:
+
+    DownLiftEquality:
+      Γ ⊢ T: A → 𝓤
+      Γ ⊢ c: A
+      Γ ⊢ x: A
+      Γ ⊢ a: T<c>
+      Γ ⊢ a: T<x>
+      ────────────
+      Γ ⊢ c = x
+
+So, if two types share constructor by some Π-constructor, share a value, their
+value parameter is equal. Take `a: [u8; 4]` as an example. If we have some
+unknown variable `x`, such that `a: [u8; x]`, we can infer that `x = 4`.
+
+This allows us to infer e.g. array length parameters in functions:
+
+```rust
+// [T; N] is a constructor, T → usize → 𝓤 (parameterize over T and you get A → 𝓤).
+fn foo<const n: usize, const l: [u32; n]>() -> [u32; n] {
+    // ^ note how l depends on n.
+    l
+}
+
+// We know n from the length of the array.
+let l = foo::<_, [1, 2, 3, 4, 5, 6]>();
+//            ^   ^^^^^^^^^^^^^^^^
+```
+
+## Structural equality
+
+Structural equality plays a key role in type checking of dependent types.
+
+Structural equality, in this case, is defined as an equivalence relation, which
+allows substitution without changing semantics.
+
+Any constant parameter must have the `structural_match` property as defined in
+[RFC #1445](https://github.com/rust-lang/rfcs/pull/1445). This property, added
+through the `#[structural_match]` attribute, essentially states that the `Eq`
+implementation is structural.
+
+Without this form of equality, substitution wouldn't be possible, and thus
+typechecking an arbitrarily value-depending type constructor would not be
+possible.
+
+## Parsing
+
+Introducing expr subgrammar in type position isn't possible without setting
+some restrictions to the possible expressions.
+
+In this case, we put restrictions to arithmetic and bitwise operations (`+-/!^`
```

Comment:
I prefer `{...}` because of the reduced verbosity, and because it is already expression syntax (if we want to use some crazy cover grammar scheme in the future).